port to FreeBSD armv6/7

MikaelUrankar opened the pull request:
Hi,
I would like to get a review on these changes please :)

I'm seeing this error on one of my board but I can't find the issue, any idea?
https://github.com/MikaelUrankar/openjdk-jdk12u/issues/1

Commit:
Import linux_arm

mkdir src/hotspot/os_cpu/bsd_arm
for file in $(ls src/hotspot/os_cpu/linux_arm/*)
do
  bsd=$(echo $file | sed 's/linux/bsd/g')
  cp ${file} ${bsd}

  sed -i '' 's/linux/bsd/g' ${bsd}
  sed -i '' 's/Linux/Bsd/g' ${bsd}
  sed -i '' 's/LINUX/BSD/g' ${bsd}
done

Commit:
Convert GNU assembler syntax to Unified Assembler Language (compatible with
clang integrated as)

error: instruction 'addlt' can not set flags, but 's' suffix specified
__asm ("addlts  r2, #32");
       ^
<inline asm>:1:2: note: instantiated into assembly here
        addlts  r2, #32
        ^

error: invalid instruction, did you mean: ldrexh, ldrh?
__asm ("ldrgeh  r4, [r0], #2");
       ^
<inline asm>:1:2: note: instantiated into assembly here
        ldrgeh  r4, [r0], #2
        ^~~~~~

error: invalid instruction, did you mean: ldrh, ldrt?
__asm ("ldrgth  r5, [r0], #2");
       ^
<inline asm>:1:2: note: instantiated into assembly here
        ldrgth  r5, [r0], #2

Unfortunately it breaks gcc

Commit:
Fix byteswap on FreeBSD

Commit:
uname returns arm on FreeBSD but we want armv6 or armv7. Use sysctl instead.

Commit:
Use __clear_cache

src/hotspot/cpu/arm/icache_arm.cpp:34:27: error: cannot initialize a parameter of type 'char *' with an lvalue of type 'address' (aka 'unsigned char *')
  __builtin___clear_cache(addr, addr + (lines << ICache::log2_line_size));
                          ^~~~
1 error generated.

Commit:
Don't use linux atomic/membar kernel helper on FreeBSD

Commit:
pass -DARM flag with clang

Commit:
clang doesn't understand externally_visible

warning: unknown attribute 'externally_visible' ignored [-Wunknown-attributes]

Commit:
Fix warning:

src/hotspot/cpu/arm/assembler_arm.hpp:187:52: warning: '&' within '|' [-Wbitwise-op-parentheses]
             (abs(_disp) & 0xf0) << 4 | abs(_disp) & 0x0f;
                                      ~ ~~~~~~~~~~~^~~~~~
src/hotspot/cpu/arm/assembler_arm.hpp:187:52: note: place parentheses around the '&' expression
to silence this warning
             (abs(_disp) & 0xf0) << 4 | abs(_disp) & 0x0f;
                                                   ^
                                        (                )

Commit:
bump cache line length to 64 bytes
The current default is too low and some jit code crashes due to stale
instruction cache

Commit:
Disable building libsaproc, it's not implemented on arm
Can't find a configure flags to disable it :(

Commit:
Do not build arm64 files on bsd-arm

Commit:
Disable some parts of os_perf_bsd.cpp. the 'struct frame' variable included
with sys/user.h clash with 'class frame' from src/hotspot/share/runtime/frame.hpp

In file included from src/hotspot/os/bsd/os_perf_bsd.cpp:37:
In file included from /usr/include/sys/user.h:38:
In file included from /usr/include/machine/pcb.h:41:
/usr/include/machine/frame.h:131:8: error: redefinition of 'frame'
struct frame {
       ^
/usr/ports/java/openjdk11/work/openjdk-jdk11u-jdk-11.0.4-11-2/src/hotspot/share/runtime/frame.hpp:55:7: note: previous definition is here
class frame {

Commit:
Add rpath for bsd-arm
The jdk libraries don't contain the ORIGIN rpath and fail to load

Commit:
port os_bsd_arm.cpp to FreeBSD

battleblow commented on make/autoconf/flags-cflags.m4:
```diff
@@ -734,6 +734,13 @@ AC_DEFUN([FLAGS_SETUP_CFLAGS_CPU_DEP],
     fi
   fi
 
+  if test "x$TOOLCHAIN_TYPE" = xclang; then
```
I assume from this that arm6/7 use clang rather than gcc?

battleblow commented on make/lib/Lib-jdk.hotspot.agent.gmk:
```diff
@@ -80,6 +80,8 @@ $(eval $(call SetupJdkLibrary, BUILD_LIBSA, \
 ))
 
-TARGETS += $(BUILD_LIBSA)
+ifneq ($(OPENJDK_TARGET_OS)-$(OPENJDK_TARGET_CPU), bsd-arm)
```
Please look at make/autoconf/jdk-options.m4 and configure an appropriate rule there to set INCLUDE_SA to false on FreeBSD/arm rather than making a change here.  The decision whether to compile the serviceability agent or not is one that should be made at configuration time.

MikaelUrankar replied:
Oh! thanks a lot. I searched libsa and saproc in the src tree and never found this!

battleblow commented on src/hotspot/cpu/arm/icache_arm.cpp:
```diff
@@ -67,7 +67,7 @@ static int icache_flush(address addr, int lines, int magic) {
 
 static int icache_flush(address addr, int lines, int magic) {
-  __builtin___clear_cache(addr, addr + (lines << ICache::log2_line_size));
+  __clear_cache(addr, addr + (lines << ICache::log2_line_size));
```
Does this break Linux and/or gcc compilation?

MikaelUrankar replied:
I don't know, maybe I should put a cast to please clang, this is the error message:
```
src/hotspot/cpu/arm/icache_arm.cpp:34:27: error: cannot initialize a parameter of type 'char *' with an lvalue of type 'address' (aka 'unsigned char *')
  __builtin___clear_cache(addr, addr + (lines << ICache::log2_line_size));
                          ^~~~
1 error generated.
```

battleblow replied:
Ok.  So the problem seems to be that GCC and Clang have different prototypes for __builtin___clear_cache.  There are a couple of ways you could deal with this.  Casting is probably the simplest of those.  Another option would potentially be to use #ifdef __GNUC__ to separate the gcc and clang versions.

MikaelUrankar replied:
FWIW aarch64 uses [__clear_cache](https://github.com/battleblow/openjdk-jdk11u/blob/bsd-port/src/hotspot/cpu/aarch64/icache_aarch64.hpp#L37)

strejda replied:
__builtin___clear_cache can be used also for clang and arm32 without problem. The only issue is that some versions of gcc declares it with void * arguments, other version of gcc and clang declares it as char *.
So cast to (char *) is safe for both variants.

battleblow commented on src/hotspot/cpu/arm/icache_arm.hpp:
```diff
@@ -32,8 +32,8 @@ class ICache : public AbstractICache {
     stub_size      = 32,                // Size of the icache flush stub in bytes
-    line_size      = BytesPerWord,      // conservative
-    log2_line_size = LogBytesPerWord    // log2(line_size)
+    log2_line_size = 6,
```
Why are we hardcoding values into a shared file?  Do the settings really not work for FreeBSD?  If so we'll need to come up with something other than this.

MikaelUrankar replied:
The value of log2_line_size is too low for FreeBSD; `LogBytesPerWord = 2` and `BytesPerWord  = 1 << LogBytesPerWord;` An instruction cache line lenght is either 32 bytes or 64 bytes (it depends of the cpu). With such a low value, the end result is a SIGILL in a perfectly supported instruction (sometime it's a sigsegv)

[LogBytesPerWord](https://github.com/battleblow/openjdk-jdk11u/blob/master/src/hotspot/share/utilities/globalDefinitions.hpp#L120)
[BytesPerWord](https://github.com/battleblow/openjdk-jdk11u/blob/master/src/hotspot/share/utilities/globalDefinitions.hpp#L126)

MikaelUrankar replied:
It works on cortex-a7, cortex-a8, cortex-a9 and cortex-a15 with `log2_line_size = 6`

Should I patch src/hotspot/share/utilities/globalDefinitions.hpp instead?

battleblow replied:
Does 8 work instead of 6?  I'm curious as to how big this has to be and how you arrived at 6.

Beyond that, my next question would be whether there is a value that would work on both Linux and FreeBSD?  If there is, then we can potentially fix up a define in this file for the line size that makes sense for both.  If not then the data should come a platform specific header.

MikaelUrankar replied:
Yes it works with 8. I put 6 because the icache line length of my cpu is 64 (2**6) and 5 was causing sigill/sigsegv. This value is used for cache maintenance with __clear_cache, putting too high value can cause sigsegv.
I don't know linux at all and I don't know why it works for them :(

battleblow replied:
I'll suggest something like this then

```
#ifdef __linux__
#define LINE_SIZE BytesPerWord
#define LOG_LINE_SIZE LogBytesPerWord
#elif _ALLBSD_SOURCE
#define LINE_SIZE BytesPerLong
#define LOG_LINE_SIZE LogBytesPerLong
#else 
#error "Platform not supported"
#endif
```

And then here use LINE_SIZE and LOG_LINE_SIZE instead of directly using either values from globals or hardcoded values.

MikaelUrankar replied:
BytesPerLong / LogBytesPerLong are still too small (I get a line_size of 8).
How about this instead:
```
--- icache_arm.hpp.orig 2019-09-02 16:55:14.234650000 +0200
+++ icache_arm.hpp      2019-09-02 16:54:57.852151000 +0200
@@ -25,6 +25,10 @@
 #ifndef CPU_ARM_VM_ICACHE_ARM_HPP
 #define CPU_ARM_VM_ICACHE_ARM_HPP
 
+#if defined(__FreeBSD__)
+#include <machine/param.h>
+#endif
+
 // Interface for updating the instruction cache.  Whenever the VM modifies
 // code, part of the processor instruction cache potentially has to be flushed.
 
@@ -32,8 +36,13 @@ class ICache : public AbstractICache {
  public:
   enum {
     stub_size      = 32,                // Size of the icache flush stub in bytes
+#if defined(__FreeBSD__)
+    log2_line_size = CACHE_LINE_SHIFT,
+    line_size      = 1 << log2_line_size
+#else
     line_size      = BytesPerWord,      // conservative
     log2_line_size = LogBytesPerWord    // log2(line_size)
+#endif
   };
 };
```
CACHE_LINE_SHIFT is defined [here](https://svnweb.freebsd.org/base/head/sys/arm/include/param.h?revision=325966&view=markup#l112)

battleblow replied:
I'm fine with something like that

strejda replied:
This is bug in FreeBSD kernel. I just committed fix  for this issue , which will be MFCed to 13.0, 13-stable and 12-stable.  - https://cgit.FreeBSD.org/src/commit/?id=ec090f4a67654fa541e6d97fd5f74d3f66c1c0d0
For 12.1 and 12.2 we can hold local patch in port tree,  I think.

battleblow commented on src/hotspot/os/bsd/os_perf_bsd.cpp:
```diff
@@ -34,7 +34,9 @@
   #include <mach/mach.h>
   #include <mach/task_info.h>
 #else
+#if !defined(__arm__)
```
Does arm not have this header?  Does it not support the sysctl to get all processes?

MikaelUrankar replied:
We have this header, the problem is the duplicate definition of the "frame" variable. We define it on armv6/7 in [frame.h](https://svnweb.freebsd.org/base/head/sys/arm/include/frame.h?revision=326258&view=markup#l133) and "frame" is also used in [frame.hpp](https://github.com/battleblow/openjdk-jdk11u/blob/master/src/hotspot/share/runtime/frame.hpp#L50)

This is the error message:
```
In file included from src/hotspot/os/bsd/os_perf_bsd.cpp:37:
In file included from /usr/include/sys/user.h:38:
In file included from /usr/include/machine/pcb.h:41:
/usr/include/machine/frame.h:131:8: error: redefinition of 'frame'
struct frame {
       ^
/usr/ports/java/openjdk11/work/openjdk-jdk11u-jdk-11.0.4-11-2/src/hotspot/share/runtime/frame.hpp:55:7: note: previous definition is here
class frame {
```

battleblow replied:
Hmm, arm must be unusual here since this doesn't happen on other architectures.  I don't see frame.hpp explicitly in os_perf_bsd.cpp, so I wonder what is pulling it in?  We may be able to avoid one or the other definitions since neither are explicitly used in the code.

MikaelUrankar replied:
I just put an `#error "mik frame.hpp" `in frame.hpp and attempted a build on amd64:
```
In file included from src/hotspot/share/precompiled/precompiled.hpp:37:         
In file included from src/hotspot/share/gc/shared/collectedHeap.hpp:36:         
In file included from src/hotspot/share/utilities/events.hpp:30:                
In file included from src/hotspot/share/runtime/thread.hpp:29:                  
In file included from src/hotspot/share/code/compiledMethod.hpp:28:             
In file included from src/hotspot/share/code/codeBlob.hpp:31:                   
src/hotspot/share/runtime/frame.hpp:37:2: error: "mik frame.hpp"                
#error "mik frame.hpp"
```

battleblow replied:
My first suggestion then is to turn off pre-compiled headers on arm6/7 since I don't think you will pull in frame.hpp in that case.

MikaelUrankar replied:
It changes nothing, codeBlob.hpp includes frame.hpp

battleblow replied:
It's pretty hacky, but it might also work to add a #define _LOCORE and forward declare trapframe just before including sys/user.h.  I can't try this without arm hardware or maybe emulation, but that might work based on https://svnweb.freebsd.org/base/head/sys/arm/include/frame.h?revision=326258&view=markup 

Really the arm headers should do a similar thing to other arches and just forward declare what they need themselves rather than including frame.h, if for nothing other than consistency.

strejda replied:
This is caused by unwanted, unnecessary header pollution in FreeBSD should be handled in slightly more elegant way - by defining  #define _MACHINE_PCB_H_ before inclusion  of sys/user.h. See https://github.com/strejda/openjdk-jdk8u/commit/ab4be68dcf7e093482ee68208a19699e216fc693.
I will fix this problem for FreeBSD 14 (head) and 13 (stable), but I think that we can have this small workaround in openjdk  bsd port.

battleblow reviewed (changes requested):
This needs a few changes.  I'm also concerned about the changes to shared files and would like some more information on those, in particular the one to icache_arm.hpp

Commit:
Revert cd1a10dde18442f2e5624a736cac3c420ae11f75
The correct fix is to disable SA for freebsd/arm in autoconf/jdk-options.m4

Commit:
Revert https://github.com/battleblow/openjdk-jdk11u/pull/74/commits/c46878657164509f0497e1c28bfa94ed37a6921b
Cast __builtin____clear_cache arguments

Commit:
Fix the following warning:
src/hotspot/os_cpu/bsd_arm/os_bsd_arm.cpp:524:43: error: format specifies type 'unsigned long' but the argument has type '__greg_t' (aka 'unsigned in
t') [-Werror,-Wformat]
  st->print_cr("  %-4s = 0x%08lx", "cpsr",uc->uc_mcontext.arm_cpsr);
                           ~~~~~          ^~~~~~~~~~~~~~~~~~~~~~~~
                           %08x

battleblow commented on src/hotspot/os_cpu/bsd_arm/bsd_arm_64.s:
```diff
@@ -0,0 +1,542 @@
+# 
```
Do we even need this file?  Aren't we only dealing with arm 32 bit hardware here?

battleblow reviewed (commented):
I'll also note that I don't have arm32 hardware to test this on, so I wanted to ask how prepared you are to maintain this and do test builds going forward?

I think it's also worth asking how many users are wanting to run Java on FreeBSD arm6/7?

MikaelUrankar commented:
> I'll also note that I don't have arm32 hardware to test this on, so I wanted to ask how prepared you are to maintain this and do test builds going forward?
> 
> I think it's also worth asking how many users are wanting to run Java on FreeBSD arm6/7?

I'm probably the only one, I'm doing this just for fun, let's close it.

strejda commented:
There are many more users  :) - for example I want to run newer java on more than 1k of custom embedded boards. Please understand that arm32 is currently in best shape for embedded application. SoCs are cheap, powerful, with many industrial interfaces (nothing like overpriced mobile based or server based SoCs in current aarch64 world.

MikaelUrankar commented:
I can't reopen the PR, I'll open a new one

edit: https://github.com/battleblow/openjdk-jdk11u/pull/103
